lib/libvfio-user.c

```diff
@@ -1519,7 +1548,18 @@ vfu_dma_read(vfu_ctx_t *vfu_ctx, dma_sg_t *sg, void *data)
     ret = vfu_ctx->tran->send_msg(vfu_ctx, msg_id, VFIO_USER_DMA_READ,
                                   &dma_send, sizeof(dma_send), NULL,
                                   dma_recv, recv_size);
-    memcpy(data, dma_recv->data, sg->length); /* FIXME no need for memcpy */
+
+    if (ret == -ENOMSG) {
+        vfu_reset_ctx(vfu_ctx, "closed");
+        ret = -ENOTCONN;
+    } else if (ret == -ECONNRESET) {
+        vfu_reset_ctx(vfu_ctx, "reset");
+        ret = -ENOTCONN;
+    } else if (ret == 0) {
+        /* FIXME no need for memcpy */
```

Comment:
Are we sure we're not missing some other `ret < 0` case? Shall we at least have an `assert`? I've looked at `tran_sock_send_iovec` and it calls `sendmsg` where more errnos can be returned.

Reply:
Not sure what you mean, we're not missing other cases, just if we get those particular two, we treat it as a reset. Others just fall through to the normal error handling.